(MODULES-4265) Detect security updates from multiple sources

nhinds opened the pull request:
The apt_security_updates fact now detects security
updates which come from more than one source
repository.

nhinds commented:
Is there anything else required to get this merged? I've got the [JIRA ticket](https://tickets.puppetlabs.com/browse/MODULES-4265) created, tests have been added, the travis build is passing, there are no merge conflicts with master.

Pinging @eputnam and @hunner as the only people who have merged pull requests into this repository in the past 2 months. Do either of you see anything that needs addressing before this is merged?

tphoney commented:
@nhinds There has been substantial change in apt since this PR was submitted. I think most of the functionality has been merged. @nhinds does it seem to be the case ?

nhinds commented:
There have been no changes to the functionality or to the test related to this bug, other than MODULES-4118 (86f654a13baec042b6d249c3f6fb875a00351f98) which set the NoLocking option and 1953cc8daac9686bfb8c2db6c76646dffe728c30 which added new code which exposes the list of packages eligible for security updates.

If `/usr/bin/apt-get -s -o Debug::NoLocking=true upgrade 2>&1` returns multiple package sources, and the `-security` source is not the last one listed, the package is still excluded.

Example from a xenial system I have checked out the current `master` branch on, where it still fails to detect the `libplist3` package as a security update:
```
$ /usr/bin/apt-get -s -o Debug::NoLocking=true upgrade
NOTE: This is only a simulation!
      apt-get needs root privileges for real execution.
      Keep also in mind that locking is deactivated,
      so don't depend on the relevance to the real current situation!
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
The following packages will be upgraded:
  libplist3 libpulse-mainloop-glib0 libpulse0 libpulsedsp pulseaudio pulseaudio-module-bluetooth
  pulseaudio-module-x11 pulseaudio-utils
8 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Inst libplist3 [1.12-3.1] (1.12-3.1ubuntu0.16.04.1 Ubuntu:16.04/xenial-security, Ubuntu:16.04/xenial-updates [amd64])
Inst libpulsedsp [1:8.0-0ubuntu3.3] (1:8.0-0ubuntu3.4 Ubuntu:16.04/xenial-updates [amd64]) []
Inst pulseaudio-utils [1:8.0-0ubuntu3.3] (1:8.0-0ubuntu3.4 Ubuntu:16.04/xenial-updates [amd64]) []
Inst pulseaudio-module-x11 [1:8.0-0ubuntu3.3] (1:8.0-0ubuntu3.4 Ubuntu:16.04/xenial-updates [amd64]) []
Inst pulseaudio-module-bluetooth [1:8.0-0ubuntu3.3] (1:8.0-0ubuntu3.4 Ubuntu:16.04/xenial-updates [amd64]) []
Inst pulseaudio [1:8.0-0ubuntu3.3] (1:8.0-0ubuntu3.4 Ubuntu:16.04/xenial-updates [amd64]) []
Inst libpulse-mainloop-glib0 [1:8.0-0ubuntu3.3] (1:8.0-0ubuntu3.4 Ubuntu:16.04/xenial-updates [amd64]) []
Inst libpulse0 [1:8.0-0ubuntu3.3] (1:8.0-0ubuntu3.4 Ubuntu:16.04/xenial-updates [amd64])
Conf libplist3 (1.12-3.1ubuntu0.16.04.1 Ubuntu:16.04/xenial-security, Ubuntu:16.04/xenial-updates [amd64])
Conf libpulse0 (1:8.0-0ubuntu3.4 Ubuntu:16.04/xenial-updates [amd64])
Conf libpulsedsp (1:8.0-0ubuntu3.4 Ubuntu:16.04/xenial-updates [amd64])
Conf pulseaudio-utils (1:8.0-0ubuntu3.4 Ubuntu:16.04/xenial-updates [amd64])
Conf pulseaudio (1:8.0-0ubuntu3.4 Ubuntu:16.04/xenial-updates [amd64])
Conf pulseaudio-module-x11 (1:8.0-0ubuntu3.4 Ubuntu:16.04/xenial-updates [amd64])
Conf pulseaudio-module-bluetooth (1:8.0-0ubuntu3.4 Ubuntu:16.04/xenial-updates [amd64])
Conf libpulse-mainloop-glib0 (1:8.0-0ubuntu3.4 Ubuntu:16.04/xenial-updates [amd64])

puppetlabs-apt$ irb
irb(main):001:0> $LOAD_PATH << './lib'
=> ["/usr/local/lib/site_ruby/2.3.0", "/usr/local/lib/x86_64-linux-gnu/site_ruby", "/usr/local/lib/site_ruby", "/usr/lib/ruby/vendor_ruby/2.3.0", "/usr/lib/x86_64-linux-gnu/ruby/vendor_ruby/2.3.0", "/usr/lib/ruby/vendor_ruby", "/usr/lib/ruby/2.3.0", "/usr/lib/x86_64-linux-gnu/ruby/2.3.0", "./lib"]
irb(main):002:0> require 'facter'
=> true
irb(main):003:0> require 'facter/apt_updates'
=> true
irb(main):004:0> Facter[:apt_security_updates].value
=> 0
irb(main):005:0> Facter[:apt_package_security_updates].value
=> []
irb(main):006:0> Facter[:apt_updates].value
=> 8
irb(main):007:0> Facter[:apt_package_updates].value
=> ["libplist3", "libpulsedsp", "pulseaudio-utils", "pulseaudio-module-x11", "pulseaudio-module-bluetooth", "pulseaudio", "libpulse-mainloop-glib0", "libpulse0"]
irb(main):008:0> 
```

The new fact `apt_package_security_updates` suffers from this same problem.

I have rebased my changes onto the current master branch and added additional tests for the `apt_package_security_updates` fact for this scenario.

eputnam commented:
Thanks @nhinds !
